ibm/resource_ibm_ipsec_vpn.go

```diff
@@ -0,0 +1,466 @@
+	}
+
+	_, err = services.GetNetworkTunnelModuleContextService(sess).
+		Id(fwID).
+		GetObject()
+
+	if err != nil {
+		if apiErr, ok := err.(sl.Error); ok && apiErr.StatusCode == 404 {
+			return false, nil
+		}
+		return false, fmt.Errorf("Error retrieving vpn information: %s", err)
+	}
+
+	return true, nil
+}
+
+func resourceIBMIPSecVPNDelete(d *schema.ResourceData, meta interface{}) error {
+	sess := meta.(ClientSession).SoftLayerSession()
+	vpnService := services.GetNetworkTunnelModuleContextService(sess)
+
+	vpnID, _ := strconv.Atoi(d.Id())
+
+	// Get billing item associated with the firewall
+	billingItem, err := vpnService.Id(vpnID).GetBillingItem()
+
+	if err != nil {
+		return fmt.Errorf("Error while looking up billing item associated with the firewall: %s", err)
+	}
+
+	if billingItem.Id == nil {
+		return fmt.Errorf("Error while looking up billing item associated with the firewall: No billing item for ID:%d", vpnID)
+	}
+
+	success, err := services.GetBillingItemService(sess).Id(*billingItem.Id).CancelService()
+	if err != nil {
+		return err
+	}
+
+	if !success {
+		return fmt.Errorf("SoftLayer reported an unsuccessful cancellation")
+	}
+
+	return nil
+}
+
+func resourceIBMIPSecVPNUpdate(d *schema.ResourceData, meta interface{}) error {
+	sess := meta.(ClientSession).SoftLayerSession()
+	vpnID, err := strconv.Atoi(d.Id())
+	var addresstranslation datatypes.Network_Tunnel_Module_Context_Address_Translation
+	if err != nil {
+		return fmt.Errorf("Not a valid ID, must be an integer: %s", err)
+	}
+
+	vpn, err := services.GetNetworkTunnelModuleContextService(sess).
+		Id(vpnID).Mask(ipsecMask).
+		GetObject()
+
+	if err != nil {
+		return fmt.Errorf("Error updating storage information: %s", err)
+	}
+	if d.HasChange("phase_one") {
+		for _, e := range d.Get("phase_one").([]interface{}) {
+			value := e.(map[string]interface{})
+			auth := value["authentication"].(string)
+			vpn.PhaseOneAuthentication = &auth
+			encryption := value["encryption"].(string)
+			vpn.PhaseOneEncryption = &encryption
+			diffiehellman := value["diffie_hellman_group"].(int)
+			vpn.PhaseOneDiffieHellmanGroup = &diffiehellman
+			keylife := value["keylife"].(int)
+			vpn.PhaseOneKeylife = &keylife
+		}
+	}
+	if d.HasChange("phase_two") {
+		for _, e := range d.Get("phase_two").([]interface{}) {
+			value := e.(map[string]interface{})
+			auth := value["authentication"].(string)
+			vpn.PhaseTwoAuthentication = &auth
+			encryption := value["encryption"].(string)
+			vpn.PhaseTwoEncryption = &encryption
+			diffiehellman := value["diffie_hellman_group"].(int)
+			vpn.PhaseTwoDiffieHellmanGroup = &diffiehellman
+			keylife := value["keylife"].(int)
+			vpn.PhaseTwoKeylife = &keylife
+		}
+	}
+	if d.HasChange("preshared_key") {
+		presharedkey := d.Get("preshared_key").(string)
+		vpn.PresharedKey = &presharedkey
+	}
+	if _, ok := d.GetOk("customer_peer_ip"); ok {
+		if d.HasChange("customer_peer_ip") {
```

Comment:
if and else loop has similar code...Do we need loop here?

Reply:
Done. Please review